Misc. changes in PAL_Random(): use getrandom(), fix O_CLOEXEC usage, etc.

lpereira opened the pull request:
There are two big changes here:

- Random data from /dev/urandom should not be messed around with, especially with a PRNG that has been initialized with little entropy (`time(NULL)`). Only use the userland PRNG if all other methods fail.
- `getrandom(2)`, introduced in Linux 3.17, is used if available. This makes it possible to gather data from the same pool `/dev/urandom` draws from, without requiring that device to exist (e.g. under a tight `chroot()`), and do not require opening a file.

lpereira committed:
Use URANDOM_DEVICE_NAME to reference /dev/urandom

URANDOM_DEVICE_NAME was defined but not used.

lpereira committed:
Actually open /dev/urandom with O_CLOEXEC

O_CLOEXEC isn't part of `mode`, but rather of the `flags` argument.

lpereira committed:
Only use mrand48() if reading random numbers from /dev/urandom fails

It's safe to assume that /dev/urandom will provide good pseudo-random
numbers.  Mucking with these numbers using another PRNG might even
deteriorate the quality of these numbers.

lpereira committed:
Use getrandom() if available to implement PAL_Random()

This system call has been added to Linux in 3.17.  The main benefit
of it is not requiring /dev/urandom to exist -- which might be the
case if running under a chroot where /dev isn't present.  It also
does not require another file descriptor.  It's otherwise equivalent
to reading from /dev/urandom (if passing the GRND_NONBLOCK flag).

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +230,154 @@ PAL_GetPALDirectoryA(
+            }
+            else if ((errno == EINTR) || (errno == EAGAIN))
+            {
+                continue;
```
Do you know what the getrandom doc means by "if the urandom source has not yet been initialized"? Does it mean that enough entropy was not accumulated or is it something else? I've seen notes that waiting for the entropy pool initialization may take non-trivial amount of time. If that applied here, I would just fail in case it would block.
The PAL_Random is not supposed to be used for any cryptographical stuff, we use it to generate Guids for internal purposes (debugger) and random numbers in superpmi jit tests. Blocking would not be desired here.

lpereira replied:
This is often just in early boot situations where there are no hardware-provided entropy sources. For systems with RDRAND (or equivalent), networked devices (so lots of interrupt activity going on), etc. this isn't much of an issue.  In any case, I've modified the code so it doesn't retry on EAGAIN, only on EINTR -- so it's closer to what the /dev/urandom function does.

janvorli commented on src/pal/src/configure.cmake:
```diff
@@ -130,6 +130,8 @@ check_function_exists(pthread_mutex_init HAS_PTHREAD_MUTEXES)
 check_function_exists(ttrace HAVE_TTRACE)
 check_function_exists(pipe2 HAVE_PIPE2)
 
+check_symbol_exists(getrandom sys/random.h HAVE_GETRANDOM)
```
We build our official builds on CentOS 7 to allow them to work on a large number of distros. CentOS 7 headers don't contain getrandom definition, although they do have __NR_getrandom defined. So we could use syscall function with that instead.

lpereira replied:
Ah, right. Yes, good suggestion. Will do this.

krwq commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +230,154 @@ PAL_GetPALDirectoryA(
-        num >>= 8;
+        for (DWORD i = 0; i < dwLength; i++)
+        {
+            long num = 0;
```
should this be outside of the loop? otherwise you are xor-ing  with 0 on 3/4 iterations

lpereira replied:
Good catch.

lpereira commented:
Thanks for the review! I've force-pushed addressing the comments.

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +234,155 @@ PAL_GetPALDirectoryA(
+
+    do
+    {
+        ssize_t res = syscall(SYS_getrandom, lpBuffer, length,
```
We are missing updating the lpBuffer, so in every loop, we end up putting the read portion of the data at the beginning of the buffer.

lpereira replied:
Done.

lpereira commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +234,156 @@ PAL_GetPALDirectoryA(
+                num = mrand48();
+            }
+
+            *(((BYTE*)lpBuffer) + i) ^= num;
```
While looking at the corefx counterpart of this code, I noticed that this is wrong: I'm trusting getrandom()/urandom here, and only using mrand48() if those didn't produce anything. So this XORs with potentially uninitialized values, which is UB territory. I'm fixing this to memcpy() in sizeof(long) chunks instead.

lpereira committed:
When falling back to mrand48(), copy -- and not XOR -- bytes to buffer

Exclusive-ORing those bytes were OK when mrand48() wasn't being used
solely as the fallback for when getrandom(2) and /dev/urandom weren't
available.  Since the buffer contents are most likely uninitialized,
XORing their contents is undefined behavior at this point; so just
write the random bytes.

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -375,14 +375,17 @@ PAL_Random(
-                num = mrand48();
-            }
+            long num = mrand48();
+            memset((char *)lpBuffer + i, &num, sizeof(num));
```
It seems you meant to use `memcpy` here instead of `memset`. Also, this code is not correct, e.g. it would overwrite `sizeof(long)-1` bytes after the end of the buffer if `dwLength` was 1.

lpereira replied:
Ouch. Yes. I don't know what I was thinking here. I rewrote the function now; shouldn't overwrite anything.

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +234,159 @@ PAL_GetPALDirectoryA(
+    return TRUE;
+}
+
+#if defined(HAVE_SYS_GETRANDOM)
```
This needs to be `#if HAVE_SYS_GETRANDOM`, cmake always define the symbol, it is defined as 0 when the test failed and 1 when it succeeded.

lpereira replied:
Done

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -40,6 +40,14 @@ Revision History:
 #include <mach-o/dyld.h>
 #endif // __APPLE__
 
+#ifdef HAVE_SYS_GETRANDOM
```
This needs to be `#if HAVE_SYS_GETRANDOM` as I have mentioned in my previous comment.

lpereira replied:
Done

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +234,162 @@ PAL_GetPALDirectoryA(
-        num >>= 8;
+        while (dwLength)
+        {
+            const auto num = mrand48();
```
Please do not use auto when it is not clear what the type is.

lpereira replied:
Done.

janvorli commented on src/pal/src/misc/miscpalapi.cpp:
```diff
@@ -226,86 +234,162 @@ PAL_GetPALDirectoryA(
+        {
+            const auto num = mrand48();
+
+            if (dwLength >= sizeof(num))
```
May I suggest a little simplification? Instead of the if / else here:
```c++
DWORD copySize = std::min(dwLength, sizeof(num));
memcpy(lpBuffer, &num, copySize);
lpBuffer = (char *)lpBuffer + copySize;
dwLength -= copySize;
```

lpereira replied:
I can do this, but memcpy() with a known size at compile time generates slightly better code (if size fits in a register, it's often just a MOV when unaligned loads are OK in that platform).

janvorli replied:
Right, that's true. Forget my suggestion then.

janvorli replied:
It would be good to look at the generated code in release build to verify that the memcpy really got replaced by assignment.

lpereira replied:
Here's a [godbolt link](https://gcc.godbolt.org/z/iMWcaq) with both versions.

jkotas commented:
> Random data from /dev/urandom should not be messed around with,

What's the problem with it? Overlaying mrand over `urandom` result won't make the output less random... . 

Originally, this was implemented as a workaround for systems with broken urandom that was returning success, but not producing random bytes.

lpereira commented:
> > Random data from /dev/urandom should not be messed around with,
> 
> What's the problem with it? Overlaying mrand over `urandom` result won't make the output less random... .
> 

If values from urandom were bad, I don't think lrand48() will help much.  For instance, musl libc masks the upper 32 bits of lrand48() -- the code that was there would "flip" bits based on its output, but only every other 4 bytes.

Admittedly, this is better than nothing.  I can reinstate the old behavior; at least we'll be able to use `getrandom()`, which has some advantages over reading from urandom.

> Originally, this was implemented as a workaround for systems with broken urandom that was returning success, but not producing random bytes.

Do you remember which system was this? I remember seeing something recently about the `RDRAND` instruction not being properly implemented on some AMD processors (causing it to produce little to no entropy after a system suspend -- so userland programs using that instruction directly weren't getting good results from it), but at least on Linux, this shouldn't be much of an issue because urandom shouldn't be providing data from a single source.

If this issue still persists upstream on any of these systems we support, we should report the bug to them.

jkotas commented:
> Do you remember which system was this? 

This was number of years ago. I am not able to find the details.

janvorli commented:
> For instance, musl libc masks the upper 32 bits of lrand48() -- the code that was there would "flip" bits based on its output, but only every other 4 bytes.

In fact, by definition (based on the doc), this function returns random numbers from 0..2^32-1, even though it returns it as long. So we should be using just 32 bits from the result. 
Btw, xoring by anything cannot make the original random sequence worse (see https://crypto.stackexchange.com/questions/48145/xor-a-set-of-random-numbers), so no matter how good or bad the results from lrand48 are, they will either improve or not change the randomness of the data.

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

janvorli commented:
@lpereira I wonder what to do with this PR now. Should we move it to the dotnet/runtime repo and finalize the discussion there? It has been stale for a couple of months.

maryamariyan commented:
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.
